src/aggregate.js

```diff
@@ -0,0 +1,230 @@
+import * as Node from './node.js'
+import { log2Ceil } from './uint64.js'
+import { indexAreaStart } from './inclusion.js'
+
+const NodeSize = BigInt(Node.Size)
+const EntrySize = Number(Index.EntrySize)
+export const MAX_CAPACITY = 2n ** BigInt(Hybrid.MAX_LOG2_LEAFS) * NodeSize
+
+/**
+ * Our default aggregate size is 32GiB. 16GiB might be an option smaller pieces
+ *
+ * but our current
+ * average rate is 30 GiB per hour
+ */
+export const DEFAULT_DEAL_SIZE = Piece.PaddedSize.from(2n ** 35n) // 32 GiB
+
+export const { PaddedSize, UnpaddedSize } = Piece
+
+/**
+ * @param {object} [options]
+ * @param {API.PaddedPieceSize} [options.size] - Size of the aggregate in
+ * (fr32 padded) bytes. If omitted default to 32 GiB which is a good
+ * default because deals 8GiB and smaller are too expensive for service
+ * providers to bother. The 16GiB deals may be an option, however our
+ * current average rate of 30 GiB per hour suggests that 32 GiB is a
+ * better default.
+ */
+export const createBuilder = ({ size = DEFAULT_DEAL_SIZE } = {}) =>
+  new AggregateBuilder({ size })
+
+/**
+ * @param {object} options
+ * @param {API.PieceInfo[]} options.pieces - Pieces to add to the aggregate
+ * @param {API.PaddedPieceSize} [options.size] - Size of the aggregate in
+ * (fr32 padded) bytes. If omitted default to 32 GiB
+ */
+export const build = ({ pieces, size = DEFAULT_DEAL_SIZE }) => {
+  const builder = createBuilder({ size })
+
+  for (const piece of pieces) {
+    builder.write(piece)
+  }
+
+  return builder.close()
+}
+
+class AggregateBuilder {
+  /**
+   * @param {object} source
+   * @param {API.PaddedPieceSize} source.size
+   * @param {API.uint64} [source.offset]
+   * @param {API.MerkleTreeNodeSource[]} [source.parts]
+   * @param {number} [source.limit]
+   */
+  constructor({
+    size,
+    limit = Index.maxIndexEntriesInDeal(size),
+    offset = 0n,
+    parts = [],
+  }) {
+    this.size = PaddedSize.from(size)
+    this.offset = offset
+    this.parts = parts
+
+    this.limit = limit
+
+    this._tree = null
+  }
+
+  /**
+   * Size of the index in bytes.
+   */
+  get indexSize() {
+    return this.limit * EntrySize
+  }
+
+  close() {
+    const { size, parts, limit, offset } = this
+    const indexStartNodes = indexAreaStart(size) / NodeSize
+
+    /** @type {API.MerkleTreeNodeSource[]} */
+    const batch = new Array(2 * parts.length)
+
+    for (const [n, part] of parts.entries()) {
+      const segment = Segment.fromSourceWithChecksum(part)
+      const node = Segment.toIndexNode(segment)
+      const index = n * 2
+      batch[index] = {
+        node: segment.root,
+        location: {
+          level: 0,
+          index: indexStartNodes + BigInt(index),
+        },
+      }
+
+      batch[index + 1] = {
+        node,
+        location: {
+          level: 0,
+          index: indexStartNodes + BigInt(index + 1),
+        },
+      }
+    }
+
+    const tree = Hybrid.create(log2Ceil(this.size / NodeSize))
```

Comment:
nit: since we deconstruct the object
```suggestion
    const tree = Hybrid.create(log2Ceil(size / NodeSize))
```